Add PIDController class to sympy.physics.control module

mohakmalviya opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes: #26163

#### Brief description of what is fixed or changed

Introduced a PIDController class in the control module to simplify the creation and manipulation of PID controllers in control systems. This class allows users to define PID controllers by specifying proportional (KP), integral (KI), and derivative (KD) gains, and automatically constructs the corresponding transfer function.

This addition enhances the control module by providing a direct way to work with PID controller.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.control
  * Introduced the PIDController class in the control module.
<!-- END RELEASE NOTES -->

mohakmalviya committed:
Add PIDController class to sympy.physics.control module

Introduced a PIDController class in the control module to simplify the creation and manipulation of PID controllers in control systems. This class allows users to define PID controllers by specifying proportional (KP), integral (KI), and derivative (KD) gains, and automatically constructs the corresponding transfer function.

- The PIDController class extends the functionality of the lti.py file within the sympy.physics.control module.
- It provides an intuitive interface for creating PID controllers, supporting both symbolic and numerical coefficients.
- This enhancement addresses the need for a more streamlined approach to working with PID controllers in symbolic control system analysis and design.

Example usage:

    from sympy.abc import s, KP, KI, KD
    from sympy.physics.control import PIDController

    pid = PIDController(KP, KI, KD, s)
    print(pid.transfer_function)

This addition enhances the control module by providing a direct way to work with PID controller.

Issue Adressed: #26163 (https://github.com/sympy/sympy/issues/26163)

mohakmalviya committed:
improved code quality

moorepants commented:
A PID class should likely subclass from transfer function so that it behaves like a transfer function.

faze-geek commented:
I updated the release notes, the description must be short and the module name is precisely `physics.control`. Feel free to make improvements to it !

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1302,50 @@ def to_expr(self):
+        return super().__new__(cls, tf)
+
+    @property
+    def transfer_function(self):
```
You'll come across `.doit()` to convert other classes to `TransferFunction`. I would like that to be added instead of a function named `transfer_function` itself.

mohakmalviya replied:
sir so do we need to create a custom .doit()?

faze-geek replied:
> sir so do we need to create a custom .doit()?

I believe we do here.

faze-geek commented:
I would suggest you give a set of robust tests. With the various behaviors and functionality, you want to support. Then we can have a look at the API design.
Then the actual implementation won't seem to be a challenge!

mohakmalviya committed:
Added PID class as subclass of transfer function

faze-geek commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -505,6 +506,24 @@ def test_Series_construction():
 
+def test_PIDController_creation_and_properties():
+    KP, KI, KD = symbols('KP KI KD', real=True)
+    s = Symbol('s')
```
Unnecessary, symbol `s` is already declared on line 23.

mohakmalviya committed:
Update: Improved code quality

mohakmalviya committed:
Update : Improved code quality

mohakmalviya committed:
Update: improved code quality(3)

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,46 @@ def to_expr(self):
             return Pow(self.den, -1, evaluate=False)
 
 
+class PIDController(TransferFunction):
```
Register the class at `sympy/core/tests/test_args.py` to correct the failing optional dependency check.

faze-geek commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -505,6 +506,23 @@ def test_Series_construction():
     raises(TypeError, lambda: Series(tf3, Matrix([1, 2, 3, 4])))
 
 
+def test_PIDController_creation_and_properties():
```
Can you test it more extensively? 
I generally like if the code goes well with MATLAB unit tests / examples - https://in.mathworks.com/help/control/ref/pid.html

faze-geek replied:
Add some unit tests from the above link and add them to your test suite.

faze-geek commented:
I have left some reviews. You can resolve the addressed ones to avoid cluttering. I will go over the correctness of the code once you add more tests.
Feel free to tag me when you are done.

mohakmalviya commented:
> I have left some reviews. You can resolve the addressed ones to avoid cluttering. I will go over the correctness of the code once you add more tests. Feel free to tag me when you are done.

sure sir thankyou!

mohakmalviya commented:
> I have left some reviews. You can resolve the addressed ones to avoid cluttering. I will go over the correctness of the code once you add more tests. Feel free to tag me when you are done.

sir i have added the class in test_args.py 
![image](https://github.com/sympy/sympy/assets/38376114/9ac153ae-6bf0-4e9d-ab6c-150e1262469e)

it is still giving me test failed 
![image](https://github.com/sympy/sympy/assets/38376114/9137838f-38c4-4b21-ab4d-4deda8a46e0e)

faze-geek commented:
> sir i have added the class in test_args.py 

I don't see an `assert` statement in the attached image. 
You need not shy from making a commit. We will investigate from there.

mohakmalviya commented:
> > sir i have added the class in test_args.py
> 
> I don't see an `assert` statement in the attached image. You need not shy from making a commit. We will investigate from there.

Sure sir

mohakmalviya committed:
Register the class at sympy/core/tests/test_args.py to correct the failing optional dependency check

faze-geek commented on sympy/core/tests/test_args.py:
```diff
@@ -5438,3 +5437,10 @@ def test_sympy__combinatorics__perm_groups__Coset():
     G = PermutationGroup([a, b])
     assert _test_args(Coset(a, G))
+
+def test_sympy__physics__control__lti__PIDController():
```
Add the function somewhere after this block, not at the end of the file. https://github.com/sympy/sympy/blob/c65c1b672167abacb02d132b3538ec8aa0ed2451/sympy/core/tests/test_args.py#L4257

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,50 @@ def to_expr(self):
+    >>> from sympy.physics.control.lti import PIDController
+    >>> KP, KI, KD = 1, 0.1, 0.01
+    >>> pid = PIDController(KP, KI, KD, s)
+    >>> pid
```
Kindly test it out well. Ensure that the `doctest` output is correct on your terminal first, then change the implementation.
```
File "/home/runner/work/sympy/sympy/sympy/physics/control/lti.py", line 1324, in sympy.physics.control.lti.PIDController
Failed example:
    pid
Expected:
    TransferFunction(K_D*s + K_P + K_I/s, 1, s)        # What you expect
Got:
    PIDController(1, 0.1, 0.01, s)                     # What sympy returns
```

mohakmalviya committed:
Added simple tests for PIDController class

mohakmalviya committed:
improved code quality

mohakmalviya commented:
@faze-geek hello sir i have added basic tests please guide me how do i improve them

mohakmalviya committed:
Updated doctests

mohakmalviya committed:
Updated code quality

faze-geek commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -505,6 +507,28 @@ def test_Series_construction():
+def test_PIDController_instance():
+    KP, KI, KD = 1, 0.1, 0.01
+    pid = PIDController(KP, KI, KD, s)
+    assert isinstance(pid, PIDController), "The object should be an instance of PIDController."
```
The error message is not required. You can remove all of them.

mohakmalviya replied:
okay sir

mohakmalviya committed:
Removed unwanted code

mohakmalviya committed:
Added tests for discrete time and named input output

mohakmalviya committed:
Modified code for failing dependencies

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,97 @@ def to_expr(self):
+
+        KP, KI, KD, Tf = sympify(KP), sympify(KI), sympify(KD), sympify(Tf)
+        numerator = KD*s**2 + KP*s + KI
+        denominator = s + Tf if Tf else s
```
These should make use of the Series and Parallel blocks.

mohakmalviya replied:
done sir

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,97 @@ def to_expr(self):
+        obj.Tf = Tf
+        obj.Ts = Ts
+        obj.InputName = InputName
+        obj.OutputName = OutputName
```
Unless these names are already introduced to TransferFunction they should not be present here. They belong on a superclass, regardless.

mohakmalviya replied:
got it sir

mohakmalviya committed:
seperated P, I, D and applied parallel block

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,33 @@ def to_expr(self):
 
 
+class PIDController:
+    def __init__(self, KP, KI, KD, s):
```
I understand that `s` is the conventional symbol for continuous time systems but use var as the variable name here

mohakmalviya replied:
sure sir

mohakmalviya committed:
replaced "s" with "var" for continuous time systems

mohakmalviya commented:
i am using series and parallel class like this - 
`P = TransferFunction(KP, 1, s)
 I = TransferFunction(KI, s, s)
 D = TransferFunction(KD * s, 1, s)
pid_controller = Parallel(P, Series(I, D)).doit()`

but it is returning me  - `-0.01*s - 0.1/s` am i doing something wrong ?? @faze-geek @moorepants

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,54 @@ def to_expr(self):
+
+        self.pid_controller = Parallel(P, I, D)
+
+    def to_PID_expr(self):
```
This has not been tested yet, kindly cover it!

mohakmalviya replied:
yes sir

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,54 @@ def to_expr(self):
 
 
+class PIDController:
+    """
+    A PID Controller class that uses Parallel to combine the P, I, and D components.
+    Parameters
```
```suggestion
    """
    A PID Controller class that uses Parallel to combine the P, I, and D components.
    
    Parameters
```

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,54 @@ def to_expr(self):
+    KD : Expr, Number
+        Derivative gain.
+    s : Symbol
+        The complex frequency variable.
+    Examples
```
```suggestion
        The complex frequency variable.
        
    Examples
```

faze-geek commented:
This is trending towards the right direction. To avoid cluttering please resolve the conversations which you are done with.

mohakmalviya commented:
> This is trending towards the right direction. To avoid cluttering please resolve the conversations which you are done with.

okay sir !

mohakmalviya committed:
Added tests for parallel in PIDController class

mohakmalviya committed:
Updated suggestion 1

Co-authored-by: Anurag Bhat <90216905+faze-geek@users.noreply.github.com>

mohakmalviya committed:
Updated suggestion 2

Co-authored-by: Anurag Bhat <90216905+faze-geek@users.noreply.github.com>

mohakmalviya committed:
Updated code quality

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,56 @@ def to_expr(self):
 
+class PIDController:
+    """
+    A PID Controller class that uses Parallel to combine the P, I, and D components.
```
```suggestion
    A PID Controller class that uses Parallel connection to combine the P, I, and D components.
```

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,56 @@ def to_expr(self):
             return Pow(self.den, -1, evaluate=False)
 
 
+class PIDController:
```
I do realize this late but why is it an independent class?
If all the behavior matches that of a transfer function, should it not be a subclass?

mohakmalviya replied:
sir i think that making it a subclass of `Transferfunction()` will not be fruitful because of the inheritance issues....when i am making it a subclass of `Transferfunction()` i am getting recursion error and according to me it means that it is breaking the functionality of `Transferfunction()` class

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,56 @@ def to_expr(self):
+
+        self.pid_controller = Parallel(P, I, D)
+
+    def to_PID_expr(self):
```
`to_PID_expr()` doesn't make much sense. What are you actually trying to return here?

mohakmalviya replied:
sir i have used `to_PID_expr()` to convert PID controller to a transfer function expression

faze-geek replied:
`to_Transfer_Function()` looks more straightforward to me.

abhiphile commented:
I think the same class should also work for the P, PI and PID controllers. If I want to use PI then, I can use it like
```
pic = PIDController(kp, ki) 
```
and it should also work. So there should be some default values of the coefficients.

faze-geek commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -505,6 +507,36 @@ def test_Series_construction():
 
 
+def test_PIDController_properties():
+    var = symbols('s')
```
You can use the `s` declared globally above instead of `var` for each test.

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -1300,6 +1300,56 @@ def to_expr(self):
+    >>> print(pid.to_PID_expr())
+    TransferFunction(s*(0.01*s + 1) + 0.1, s, s)
+    """
+    def __init__(self, KP, KI, KD, var):
```
I understand your comment on default values, but can you give an example here. What would the coefficient be ?

faze-geek replied:
@abhiphile

abhiphile replied:
The default value of coefficients for PI is KD = 0 because the resultant transfer function will become `KP + KI/S`.
 For the `PD` controller `, KP+S*KD` is the resultant transfer function with KI = 0, and for the `P` type controller, we have transfer function `Kp` having `Kp = 0` and `Kd = 0`. So the default value will be  Zero here for all.

faze-geek replied:
Okay I get your suggestion. This must be handled. 

Maybe instead of directly giving 0 as default coefficient we can have something like this for flexibility-
```
def __init__(self, KP=None, KI=None, KD=None, var):
    if KP is None: KP = 0
    if KI is None: KI = 0
    if KD is None: KD = 0
    // check if the inputs are symbolic expressions or numbers in the first place
```
But this can be looked into later.

mohakmalviya committed:
Added `PIDController` class as subclass of `TransferFunction`, removed `to_PID_expr()`

mohakmalviya committed:
removed doctest error

Added author in .mailmap

Updated .mailmap(1)

updated .mailmap and test_args for failing dependencies

updated code quality

Added tests for failing dependencies in test_args.py

Updated .mailmap(2)

Added author in .mailmap

Updated .mailmap(1)

updated .mailmap and test_args for failing dependencies

updated code quality

Added tests for failing dependencies in test_args.py

Updated .mailmap(2)

faze-geek commented:
Hi, can you fix all failing checks including `test / authors` before I review it. Feel free to mark it ready for review then.

mohakmalviya commented:
> Hi, can you fix all failing checks including `test / authors` before I review it. Feel free to mark it ready for review then.

sir i don't know why these tests are failing....i ran them on my local pc they wll worked fine

mohakmalviya committed:
recorded .mailmap file

mohakmalviya commented:
@faze-geek sir @abhiphile wants to work on this issue so should i close this PR?

faze-geek commented:
> @faze-geek sir @abhiphile wants to work on this issue so should i close this PR?

Anyone is free to work on any issue in SymPy. My permission is not required. !

mohakmalviya commented:
> > @faze-geek sir @abhiphile wants to work on this issue so should i close this PR?
> 
> Anyone is free to work on any issue in SymPy. My permission is not required.

okay sir :+1: once the changes are reviewed he will continue to work from here

faze-geek commented:
Go ahead and open a new PR @abhiphile. Make sure all the reviews given here are already addressed in that PR. Closing it in favor of that.
